src/Sylius/Bundle/CoreBundle/OrderProcessing/OrderTaxesApplicator.php

```diff
@@ -0,0 +1,176 @@
+        DefaultTaxZoneProviderInterface $defaultTaxZoneProvider,
+        FactoryInterface $adjustmentFactory,
+        IntegerDistributorInterface $integerDistributor,
+        TaxRateResolverInterface $taxRateResolver,
+        ZoneMatcherInterface $zoneMatcher
+    ) {
+        $this->taxCalculator = $taxCalculator;
+        $this->defaultTaxZoneProvider = $defaultTaxZoneProvider;
+        $this->adjustmentFactory = $adjustmentFactory;
+        $this->integerDistributor = $integerDistributor;
+        $this->taxRateResolver = $taxRateResolver;
+        $this->zoneMatcher = $zoneMatcher;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function apply(OrderInterface $order)
+    {
+        $this->clearTaxes($order);
+        if ($order->isEmpty()) {
+            return;
+        }
+
+        $zone = $this->provideTaxZone($order->getShippingAddress());
+
+        if (null === $zone) {
+            return;
+        }
+
+        $this->processTaxes($order, $zone);
+    }
+
+    /**
+     * @param OrderInterface $order
+     * @param ZoneInterface $zone
+     */
+    protected function processTaxes(OrderInterface $order, ZoneInterface $zone)
+    {
+        foreach ($order->getItems() as $item) {
+            $rate = $this->taxRateResolver->resolve($item->getProduct(), array('zone' => $zone));
+
+            if (null === $rate) {
+                continue;
+            }
+
+            $percentageAmount = $rate->getAmountAsPercentage();
+            $totalTaxAmount = $this->taxCalculator->calculate($item->getTotal(), $rate);
+            $label = sprintf('%s (%s%%)', $rate->getName(), (float) $percentageAmount);
+
+            $splitTaxes = $this->integerDistributor->distribute($item->getQuantity(), $totalTaxAmount);
+
+            foreach ($splitTaxes as $key => $tax) {
+                $this->addAdjustment($item->getUnits()->get($key), $tax, $label, $rate->isIncludedInPrice());
+            }
+        }
+    }
+
+    /**
+     * @param OrderItemUnitInterface $unit
+     * @param int $amount
+     * @param string $label
+     * @param bool $included
+     */
+    protected function addAdjustment(OrderItemUnitInterface $unit, $amount, $label, $included)
+    {
+        $adjustment = $this->adjustmentFactory->createNew();
+        $adjustment->setType(AdjustmentInterface::TAX_ADJUSTMENT);
+        $adjustment->setAmount($amount);
+        $adjustment->setDescription($label);
+        $adjustment->setNeutral($included);
+
+        $unit->addAdjustment($adjustment);
+    }
+
+    /**
+     * @param AddressInterface|null $shippingAddress
+     *
+     * @return ZoneInterface|null
+     */
+    private function provideTaxZone(AddressInterface $shippingAddress = null)
+    {
+        $zone = null;
+        if (null !== $shippingAddress) {
+            $zone = $this->zoneMatcher->match($shippingAddress);
+        }
+
+        if (null === $zone) {
+            $zone = $this->defaultTaxZoneProvider->provide();
+        }
```

Comment:
Not exactly. ``ZoneMatcher::match()`` could still return ``null`` and if it would, we should try to provide it with ``$defautTaxZoneProvider``